File: ger/TTP/SV4/S3/3740.py
def nxt(i): return 0 if i == n-1 else i+1

def prev(i): return n-1 if i == 0 else i-1

for t in xrange(input()):
	
	flipped = 0
	line = [int(i) for i in raw_input().split()]
	n = line[0]

	line = line[1:]
	one = 0
	for i in xrange(n):
		if line[i] == 1:
			if line[prev(i)] == 2:
				one = prev(i)
			elif line[nxt(i)] == n:
				one = nxt(i)
			else:
				one = i
	i = one
	true = 1
	while True:
		if line[i] != true: flipped += 1
		i = nxt(i)
		true += 1
		if i == one: break

	print flipped

File: ger/TTP/SV4/S3/3740.cpp
#include <iostream>
using namespace std;

int t, n, flipped, one;

int nxt(int i){return i == n-1 ? 0 : i+1;}

int prev(int i){return i == 0 ? n-1 : i-1;}


int main(){
	ios_base::sync_with_stdio(0), cin.tie(0), cout.tie(0);
	int line[1001];

	scanf("%d", &t);
	while(t--){
		flipped = 0;
		one = 0;
		cin >> n;
		for(int i = 0; i < n; i++) cin >> line[i];
		for(int i = 0; i < n; i++){
			if (line[i] == 1){
				if (line[prev(i)] == 2){
					one = prev(i);
				}else if (line[nxt(i)] == n){
					one = nxt(i);
				}else{
					one = i;
				}
			}
		}
		int i = one;
		int tr = 1;
		while(1){
			if(line[i] != tr) flipped += 1;
			i = nxt(i);
			tr += 1;
			if(i == one) break;
		}
		cout << flipped << "\n";
	}
}